File: tests/fixtures/sample.php
<?php
// sample.php — fixture file for timebomb scanner tests.
//
// Annotation inventory (hardcoded dates, never relative to today):
//   Expired        (2018–2021): 4
//   Expiring-soon  (2025-06):   1
//   Future / OK    (2088/2099): 2

declare(strict_types=1);

namespace Sample;

use InvalidArgumentException;
use RuntimeException;

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

class AppException extends RuntimeException
{
    public function __construct(
        private readonly string $code,
        string $message,
        private readonly int $statusCode = 500,
        private readonly mixed $details  = null,
    ) {
        parent::__construct($message);
    }

    public function getCode(): string  { return $this->code; }
    public function getStatus(): int   { return $this->statusCode; }
    public function getDetails(): mixed { return $this->details; }
}

class ValidationException extends AppException
{
    public function __construct(string $message, mixed $details = null)
    {
        parent::__construct('VALIDATION_ERROR', $message, 422, $details);
    }
}

class NotFoundException extends AppException
{
    public function __construct(string $resource)
    {
        parent::__construct('NOT_FOUND', "{$resource} not found", 404);
    }
}

class ConflictException extends AppException
{
    public function __construct(string $message)
    {
        parent::__construct('CONFLICT', $message, 409);
    }
}

// ---------------------------------------------------------------------------
// Result type
// ---------------------------------------------------------------------------

final class Result
{
    private function __construct(
        private readonly bool  $ok,
        private readonly mixed $value,
        private readonly ?string $error,
    ) {}

    public static function ok(mixed $value): self  { return new self(true, $value, null); }
    public static function err(string $error): self { return new self(false, null, $error); }

    public function isOk(): bool    { return $this->ok; }
    public function unwrap(): mixed
    {
        if (!$this->ok) throw new RuntimeException($this->error);
        return $this->value;
    }
    public function unwrapOr(mixed $fallback): mixed { return $this->ok ? $this->value : $fallback; }
    public function map(callable $fn): self
    {
        return $this->ok ? self::ok($fn($this->value)) : $this;
    }

    public static function try(callable $fn): self
    {
        try { return self::ok($fn()); } catch (\Throwable $e) { return self::err($e->getMessage()); }
    }
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

readonly class ValidationFailure
{
    public function __construct(
        public string  $field,
        public string  $message,
        public ?string $value = null,
    ) {}
}

// TODO[2020-02-15]: replace with symfony/validator or respect/validation
class Validator
{
    private const EMAIL_RE = '/^[^\s@]+@[^\s@]+\.[^\s@]+$/';

    public static function required(string $field, mixed $value): ?ValidationFailure
    {
        if ($value === null || $value === '') {
            return new ValidationFailure($field, 'is required');
        }
        return null;
    }

    public static function email(string $field, string $value): ?ValidationFailure
    {
        if (!preg_match(self::EMAIL_RE, $value)) {
            return new ValidationFailure($field, 'must be a valid email address', $value);
        }
        return null;
    }

    public static function minLength(string $field, string $value, int $min): ?ValidationFailure
    {
        if (mb_strlen($value) < $min) {
            return new ValidationFailure($field, "must be at least {$min} characters", $value);
        }
        return null;
    }

    public static function maxLength(string $field, string $value, int $max): ?ValidationFailure
    {
        if (mb_strlen($value) > $max) {
            return new ValidationFailure($field, "must be at most {$max} characters", $value);
        }
        return null;
    }

    public static function collect(ValidationFailure|null ...$failures): array
    {
        return array_values(array_filter($failures));
    }
}

// ---------------------------------------------------------------------------
// Config
// ---------------------------------------------------------------------------

readonly class Config
{
    public function __construct(
        public string $host,
        public int    $port,
        public string $dsn,
        public string $jwtSecret,
        public int    $jwtExpiry,
        public int    $cacheTtl,
        public int    $rateMax,
        public int    $rateWindow,
        public string $env,
    ) {}

    public static function fromEnv(): self
    {
        return new self(
            host:       $_ENV['HOST']          ?? '0.0.0.0',
            port:       (int) ($_ENV['PORT']   ?? 3000),
            dsn:        $_ENV['DATABASE_URL']  ?? 'pgsql:host=localhost;dbname=app',
            jwtSecret:  $_ENV['JWT_SECRET']    ?? 'change-me',
            jwtExpiry:  (int) ($_ENV['JWT_EXPIRY']   ?? 3600),
            cacheTtl:   (int) ($_ENV['CACHE_TTL']    ?? 300),
            rateMax:    (int) ($_ENV['RATE_MAX']      ?? 100),
            rateWindow: (int) ($_ENV['RATE_WINDOW']   ?? 60),
            env:        $_ENV['APP_ENV'] ?? 'development',
        );
    }

    public function isProduction(): bool { return $this->env === 'production'; }
}

// ---------------------------------------------------------------------------
// Cache
// ---------------------------------------------------------------------------

// HACK[2018-10-01]: APCu/array cache; replace with Predis before production
class Cache
{
    private array $store = [];

    public function get(string $key): mixed
    {
        $entry = $this->store[$key] ?? null;
        if ($entry === null || $entry['expires_at'] <= time()) {
            unset($this->store[$key]);
            return null;
        }
        return $entry['value'];
    }

    public function set(string $key, mixed $value, int $ttlSeconds): void
    {
        $this->store[$key] = ['value' => $value, 'expires_at' => time() + $ttlSeconds];
    }

    public function delete(string $key): void
    {
        unset($this->store[$key]);
    }

    public function getOrSet(string $key, int $ttl, callable $fn): mixed
    {
        $cached = $this->get($key);
        if ($cached !== null) return $cached;
        $value = $fn();
        $this->set($key, $value, $ttl);
        return $value;
    }

    public function cleanup(): void
    {
        $now = time();
        foreach ($this->store as $key => $entry) {
            if ($entry['expires_at'] <= $now) unset($this->store[$key]);
        }
    }
}

// ---------------------------------------------------------------------------
// Rate limiter
// ---------------------------------------------------------------------------

class RateLimiter
{
    private array $store = [];

    public function __construct(
        private readonly int $windowSec,
        private readonly int $maxRequests,
    ) {}

    public function check(string $key): array
    {
        $now   = time();
        $entry = $this->store[$key] ?? null;

        if ($entry === null || $entry['reset_at'] <= $now) {
            $entry = ['count' => 0, 'reset_at' => $now + $this->windowSec];
        }

        $entry['count']++;
        $this->store[$key] = $entry;

        $remaining  = max(0, $this->maxRequests - $entry['count']);
        $allowed    = $entry['count'] <= $this->maxRequests;
        $retryAfter = $allowed ? 0 : $entry['reset_at'] - $now;

        return ['allowed' => $allowed, 'remaining' => $remaining, 'retry_after' => $retryAfter];
    }
}

// ---------------------------------------------------------------------------
// Feature flags
// ---------------------------------------------------------------------------

// TODO[2099-06-15][platform]: wire to Unleash or LaunchDarkly PHP SDK
class FeatureFlagService
{
    private array $flags = [];

    public function define(string $name, bool $enabled, int $rollout = 100, array $allowlist = []): void
    {
        $this->flags[$name] = ['enabled' => $enabled, 'rollout' => $rollout, 'allowlist' => $allowlist];
    }

    public function isEnabled(string $name, ?string $userId = null): bool
    {
        $flag = $this->flags[$name] ?? null;
        if ($flag === null || !$flag['enabled']) return false;
        if ($userId !== null && in_array($userId, $flag['allowlist'], true)) return true;
        if ($flag['rollout'] >= 100) return true;
        if ($flag['rollout'] <= 0)  return false;

        $hash   = md5("{$name}:" . ($userId ?? 'anon'));
        $bucket = hexdec(substr($hash, 0, 8)) % 100;
        return $bucket < $flag['rollout'];
    }
}

// ---------------------------------------------------------------------------
// Circuit breaker
// ---------------------------------------------------------------------------

// FIXME[2021-04-01]: no distributed state; breaks with multiple PHP-FPM workers
class CircuitBreaker
{
    private string $state       = 'closed';
    private int    $failures    = 0;
    private int    $lastFailure = 0;

    public function __construct(
        private readonly int $threshold,
        private readonly int $resetTimeoutSec,
    ) {}

    public function call(callable $fn): mixed
    {
        if ($this->state === 'open') {
            if (time() - $this->lastFailure >= $this->resetTimeoutSec) {
                $this->state = 'half-open';
            } else {
                throw new AppException('CIRCUIT_OPEN', 'Service temporarily unavailable', 503);
            }
        }

        try {
            $result        = $fn();
            $this->failures = 0;
            $this->state   = 'closed';
            return $result;
        } catch (\Throwable $e) {
            $this->failures++;
            $this->lastFailure = time();
            if ($this->failures >= $this->threshold) $this->state = 'open';
            throw $e;
        }
    }

    public function getState(): string { return $this->state; }
}

// ---------------------------------------------------------------------------
// Pagination
// ---------------------------------------------------------------------------

readonly class Page
{
    public bool $hasNext;
    public bool $hasPrev;

    public function __construct(
        public array $items,
        public int   $total,
        public int   $pageNum,
        public int   $pageSize,
    ) {
        $this->hasNext = ($pageNum - 1) * $pageSize + count($items) < $total;
        $this->hasPrev = $pageNum > 1;
    }

    public static function of(array $source, int $pageNum, int $pageSize): self
    {
        $offset = ($pageNum - 1) * $pageSize;
        $items  = array_slice($source, $offset, $pageSize);
        return new self($items, count($source), $pageNum, $pageSize);
    }
}

// ---------------------------------------------------------------------------
// Utilities
// ---------------------------------------------------------------------------

// FIXME[2025-06-08]: slugify does not handle Unicode letters; add intl extension check
function slugify(string $text): string
{
    $text = mb_strtolower($text);
    $text = preg_replace('/[^a-z0-9\s-]/', '', $text);
    $text = preg_replace('/[\s-]+/', '-', $text);
    return trim($text, '-');
}

function maskEmail(string $email): string
{
    [$local, $domain] = explode('@', $email, 2) + [1 => null];
    if ($domain === null) return $email;
    $visible = mb_strlen($local) > 2 ? mb_substr($local, 0, 2) : mb_substr($local, 0, 1);
    $stars   = str_repeat('*', max(1, mb_strlen($local) - 2));
    return "{$visible}{$stars}@{$domain}";
}

function truncate(string $s, int $maxLen, string $suffix = '…'): string
{
    return mb_strlen($s) <= $maxLen ? $s : mb_substr($s, 0, $maxLen - mb_strlen($suffix)) . $suffix;
}

function formatBytes(int $bytes): string
{
    $units = ['B', 'KB', 'MB', 'GB', 'TB'];
    $i = 0;
    $v = (float) $bytes;
    while ($v >= 1024 && $i < count($units) - 1) { $v /= 1024; $i++; }
    return round($v, 2) . ' ' . $units[$i];
}

function formatDuration(int $ms): string
{
    if ($ms < 1000)   return "{$ms}ms";
    if ($ms < 60_000) return round($ms / 1000, 1) . 's';
    return floor($ms / 60_000) . 'm ' . floor(($ms % 60_000) / 1000) . 's';
}

function chunk(array $arr, int $size): array
{
    return array_chunk($arr, $size);
}

function groupBy(array $arr, callable $keyFn): array
{
    $result = [];
    foreach ($arr as $item) {
        $result[$keyFn($item)][] = $item;
    }
    return $result;
}

File: tests/fixtures/sample.java
// sample.java — fixture file for timebomb scanner tests.
//
// Annotation inventory (hardcoded dates, never relative to today):
//   Expired        (2018–2021): 4
//   Expiring-soon  (2025-06):   1
//   Future / OK    (2088/2099): 2

package fixture;

import java.security.MessageDigest;
import java.time.Duration;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.*;
import java.util.regex.Pattern;
import java.util.stream.*;

// ---------------------------------------------------------------------------
// Errors
// ---------------------------------------------------------------------------

class AppException extends RuntimeException {
    private final String code;
    private final int    statusCode;
    private final Object details;

    AppException(String code, String message, int statusCode, Object details) {
        super(message);
        this.code       = code;
        this.statusCode = statusCode;
        this.details    = details;
    }

    String getCode()       { return code; }
    int    getStatusCode() { return statusCode; }
    Object getDetails()    { return details; }
}

class ValidationException extends AppException {
    ValidationException(String message, Object details) {
        super("VALIDATION_ERROR", message, 422, details);
    }
}

class NotFoundException extends AppException {
    NotFoundException(String resource) {
        super("NOT_FOUND", resource + " not found", 404, null);
    }
}

class ConflictException extends AppException {
    ConflictException(String message) {
        super("CONFLICT", message, 409, null);
    }
}

// ---------------------------------------------------------------------------
// Result type
// ---------------------------------------------------------------------------

sealed interface Result<T> permits Result.Ok, Result.Err {
    record Ok<T>(T value)       implements Result<T> {}
    record Err<T>(String reason) implements Result<T> {}

    default boolean isOk()  { return this instanceof Ok<T>; }
    default T       unwrap() {
        if (this instanceof Ok<T> ok) return ok.value();
        throw new NoSuchElementException(((Err<T>) this).reason());
    }
    default T unwrapOr(T fallback) {
        return this instanceof Ok<T> ok ? ok.value() : fallback;
    }
    default <U> Result<U> map(Function<T, U> fn) {
        return this instanceof Ok<T> ok ? new Ok<>(fn.apply(ok.value())) : new Err<>(((Err<T>) this).reason());
    }

    static <T> Result<T> ok(T value)        { return new Ok<>(value); }
    static <T> Result<T> err(String reason)  { return new Err<>(reason); }
    static <T> Result<T> tryGet(Supplier<T> fn) {
        try { return ok(fn.get()); } catch (Exception e) { return err(e.getMessage()); }
    }
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

record ValidationFailure(String field, String message, String value) {}

// TODO[2021-01-10]: replace hand-rolled validators with Bean Validation (JSR-380)
class Validator {
    private static final Pattern EMAIL = Pattern.compile("^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$");

    static ValidationFailure required(String field, String value) {
        return (value == null || value.isBlank())
            ? new ValidationFailure(field, "is required", null)
            : null;
    }

    static ValidationFailure email(String field, String value) {
        return (value == null || !EMAIL.matcher(value).matches())
            ? new ValidationFailure(field, "must be a valid email address", value)
            : null;
    }

    static ValidationFailure minLength(String field, String value, int min) {
        return (value == null || value.length() < min)
            ? new ValidationFailure(field, "must be at least " + min + " characters", value)
            : null;
    }

    static ValidationFailure maxLength(String field, String value, int max) {
        return (value == null || value.length() > max)
            ? new ValidationFailure(field, "must be at most " + max + " characters", value)
            : null;
    }

    @SafeVarargs
    static List<ValidationFailure> collect(ValidationFailure... failures) {
        return Arrays.stream(failures).filter(Objects::nonNull).collect(Collectors.toList());
    }
}

// ---------------------------------------------------------------------------
// Config
// ---------------------------------------------------------------------------

record Config(
    String host,
    int    port,
    String dbUrl,
    String jwtSecret,
    int    jwtExpirySec,
    int    cacheTtlSec,
    int    rateMaxRequests,
    int    rateWindowSec,
    String env
) {
    static Config fromEnv() {
        return new Config(
            getenv("HOST", "0.0.0.0"),
            Integer.parseInt(getenv("PORT", "3000")),
            getenv("DB_URL", "jdbc:postgresql://localhost/app"),
            getenv("JWT_SECRET", "change-me"),
            Integer.parseInt(getenv("JWT_EXPIRY", "3600")),
            Integer.parseInt(getenv("CACHE_TTL", "300")),
            Integer.parseInt(getenv("RATE_MAX", "100")),
            Integer.parseInt(getenv("RATE_WINDOW", "60")),
            getenv("APP_ENV", "development"));
    }

    boolean isProduction() { return "production".equals(env); }

    private static String getenv(String key, String fallback) {
        String v = System.getenv(key);
        return (v != null && !v.isBlank()) ? v : fallback;
    }
}

// ---------------------------------------------------------------------------
// Cache
// ---------------------------------------------------------------------------

// HACK[2019-09-01]: ConcurrentHashMap stand-in; replace with Caffeine or Redis
class Cache<K, V> {
    private record Entry<V>(V value, Instant expiresAt) {}

    private final ConcurrentHashMap<K, Entry<V>> store = new ConcurrentHashMap<>();

    Optional<V> get(K key) {
        var entry = store.get(key);
        if (entry == null || Instant.now().isAfter(entry.expiresAt())) {
            store.remove(key);
            return Optional.empty();
        }
        return Optional.of(entry.value());
    }

    void set(K key, V value, Duration ttl) {
        store.put(key, new Entry<>(value, Instant.now().plus(ttl)));
    }

    void delete(K key) { store.remove(key); }

    V getOrSet(K key, Duration ttl, Supplier<V> factory) {
        return get(key).orElseGet(() -> {
            V v = factory.get();
            set(key, v, ttl);
            return v;
        });
    }

    void cleanup() {
        Instant now = Instant.now();
        store.entrySet().removeIf(e -> now.isAfter(e.getValue().expiresAt()));
    }
}

// ---------------------------------------------------------------------------
// Rate limiter
// ---------------------------------------------------------------------------

class RateLimiter {
    private record Entry(int count, Instant resetAt) {}

    private final ConcurrentHashMap<String, Entry> store = new ConcurrentHashMap<>();
    private final int windowSec, maxRequests;

    RateLimiter(int windowSec, int maxRequests) {
        this.windowSec   = windowSec;
        this.maxRequests = maxRequests;
    }

    record CheckResult(boolean allowed, int remaining, int retryAfterSec) {}

    CheckResult check(String key) {
        Instant now = Instant.now();
        Entry entry = store.compute(key, (k, e) -> {
            if (e == null || now.isAfter(e.resetAt())) return new Entry(1, now.plusSeconds(windowSec));
            return new Entry(e.count() + 1, e.resetAt());
        });
        int remaining   = Math.max(0, maxRequests - entry.count());
        boolean allowed = entry.count() <= maxRequests;
        int retryAfter  = allowed ? 0 : (int) Duration.between(now, entry.resetAt()).getSeconds();
        return new CheckResult(allowed, remaining, retryAfter);
    }
}

// ---------------------------------------------------------------------------
// Pagination
// ---------------------------------------------------------------------------

record Page<T>(List<T> items, int total, int pageNum, int pageSize) {
    boolean hasNext() { return (pageNum - 1) * pageSize + items.size() < total; }
    boolean hasPrev() { return pageNum > 1; }

    static <T> Page<T> of(List<T> source, int pageNum, int pageSize) {
        int offset = (pageNum - 1) * pageSize;
        var items  = source.stream().skip(offset).limit(pageSize).collect(Collectors.toList());
        return new Page<>(items, source.size(), pageNum, pageSize);
    }
}

// ---------------------------------------------------------------------------
// Feature flags
// ---------------------------------------------------------------------------

// TODO[2099-11-01][platform]: replace with LaunchDarkly Java SDK
class FeatureFlagService {
    private record Flag(boolean enabled, int rolloutPercent, Set<String> allowlist) {}
    private final ConcurrentHashMap<String, Flag> flags = new ConcurrentHashMap<>();

    void define(String name, boolean enabled, int rollout, String... allowlist) {
        flags.put(name, new Flag(enabled, rollout, Set.of(allowlist)));
    }

    boolean isEnabled(String name, String userId) {
        Flag flag = flags.get(name);
        if (flag == null || !flag.enabled()) return false;
        if (userId != null && flag.allowlist().contains(userId)) return true;
        if (flag.rolloutPercent() >= 100) return true;
        try {
            var md5    = MessageDigest.getInstance("MD5");
            var hash   = md5.digest((name + ":" + (userId != null ? userId : "anon")).getBytes());
            int bucket = Math.abs(hash[0] & 0xFF) % 100;
            return bucket < flag.rolloutPercent();
        } catch (Exception e) { return false; }
    }
}

// ---------------------------------------------------------------------------
// Circuit breaker
// ---------------------------------------------------------------------------

// FIXME[2020-07-15]: open state does not honour half-open probe limit
class CircuitBreaker {
    private enum State { CLOSED, OPEN, HALF_OPEN }

    private volatile State   state       = State.CLOSED;
    private volatile int     failures    = 0;
    private volatile Instant lastFailure = Instant.MIN;
    private final    int     threshold;
    private final    Duration resetTimeout;

    CircuitBreaker(int threshold, Duration resetTimeout) {
        this.threshold    = threshold;
        this.resetTimeout = resetTimeout;
    }

    <T> T call(Supplier<T> fn) {
        synchronized (this) {
            if (state == State.OPEN) {
                if (Duration.between(lastFailure, Instant.now()).compareTo(resetTimeout) >= 0)
                    state = State.HALF_OPEN;
                else
                    throw new AppException("CIRCUIT_OPEN", "Service temporarily unavailable", 503, null);
            }
        }
        try {
            T result = fn.get();
            synchronized (this) { failures = 0; state = State.CLOSED; }
            return result;
        } catch (Exception e) {
            synchronized (this) {
                failures++;
                lastFailure = Instant.now();
                if (failures >= threshold) state = State.OPEN;
            }
            throw e;
        }
    }

    String getState() { return state.toString(); }
}

// ---------------------------------------------------------------------------
// Metrics
// ---------------------------------------------------------------------------

class Counter {
    private final AtomicLong value = new AtomicLong();
    void inc()          { value.incrementAndGet(); }
    void add(long n)    { value.addAndGet(n); }
    long read()         { return value.get(); }
    void reset()        { value.set(0); }
}

class Histogram {
    // TODO[2025-06-10]: wire to Micrometer and export to Prometheus
    private final List<Double> samples = Collections.synchronizedList(new ArrayList<>());

    void observe(double value) { samples.add(value); }

    double percentile(double p) {
        if (samples.isEmpty()) return 0;
        List<Double> sorted = new ArrayList<>(samples);
        Collections.sort(sorted);
        int idx = (int) Math.ceil(p / 100.0 * sorted.size()) - 1;
        return sorted.get(Math.max(0, idx));
    }
    void reset() { samples.clear(); }
}

// ---------------------------------------------------------------------------
// Utilities
// ---------------------------------------------------------------------------

class StringUtils {
    private static final Pattern NON_ALNUM = Pattern.compile("[^a-z0-9]+");

    static String slugify(String text) {
        return NON_ALNUM.matcher(text.toLowerCase()).replaceAll("-").replaceAll("^-|-$", "");
    }

    static String maskEmail(String email) {
        String[] parts = email.split("@", 2);
        if (parts.length != 2) return email;
        String local   = parts[0];
        String visible = local.length() > 2 ? local.substring(0, 2) : local.substring(0, 1);
        String stars   = "*".repeat(Math.max(1, local.length() - 2));
        return visible + stars + "@" + parts[1];
    }

    static String truncate(String s, int maxLen, String suffix) {
        return s.length() <= maxLen ? s : s.substring(0, maxLen - suffix.length()) + suffix;
    }

    static String formatBytes(long bytes) {
        String[] units = {"B", "KB", "MB", "GB", "TB"};
        double v = bytes;
        int i = 0;
        while (v >= 1024 && i < units.length - 1) { v /= 1024; i++; }
        return String.format("%.2f %s", v, units[i]);
    }

    static String formatDuration(long ms) {
        if (ms < 1000)   return ms + "ms";
        if (ms < 60_000) return String.format("%.1fs", ms / 1000.0);
        return (ms / 60_000) + "m " + (ms % 60_000 / 1000) + "s";
    }
}
